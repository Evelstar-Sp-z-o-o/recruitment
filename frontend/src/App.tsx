import { FC } from 'react';

import Root from '@/src/pages/Root';
import AppProviders from '@/src/utils/providers/AppProviders';

import './styles/main.scss';

const App: FC = () => {
  return (
    <AppProviders>
      <Root />
    </AppProviders>
  );
};

export default App;
