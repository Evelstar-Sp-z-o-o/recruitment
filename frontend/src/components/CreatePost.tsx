import { FC, useState, FormEvent } from 'react';
import { useDispatch } from 'react-redux';

import AddAPhotoTwoToneIcon from '@mui/icons-material/AddAPhotoTwoTone';
import AddPhotoAlternateTwoToneIcon from '@mui/icons-material/AddPhotoAlternateTwoTone';
import AttachFileIcon from '@mui/icons-material/AttachFile';
import { Avatar, Box, Button, Card, Grid, TextField } from '@mui/material';

import { createPost } from '../redux/actions/posts';
import { PostData } from '../types';

const CreatePost: FC = () => {
  const dispatch = useDispatch();

  const [post, setPost] = useState<PostData>({
    body: '',
    author: '',
    created: Date.now() / 1000,
    edited: Date.now() / 1000,
    postId: '',
  });

  const [errors, setErrors] = useState({
    author: '',
    body: '',
  });

  const validate = () => {
    let isValid = true;
    const newErrors = { author: '', body: '' };

    if (post.author.trim() === '') {
      newErrors.author = 'Author name is required';
      isValid = false;
    }
    if (post.body.trim() === '') {
      newErrors.body = 'Message is required';
      isValid = false;
    }

    setErrors(newErrors);
    return isValid;
  };

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    if (validate()) {
      dispatch(createPost({ id: 0, data: post }));
      setPost({
        body: '',
        author: '',
        created: Date.now() / 1000,
        edited: Date.now() / 1000,
        postId: generateRandomId(7),
      });
      setErrors({ author: '', body: '' });
    }
  };

  const generateRandomId = (length) => {
    const characters = 'abcdefghijklmnopqrstuvwxyz0123456789';
    let result = '';
    const charactersLength = characters.length;
    for (let i = 0; i < length; i++) {
      result += characters.charAt(Math.floor(Math.random() * charactersLength));
    }
    return result;
  };

  return (
    <Box sx={{ display: 'flex', alignItems: 'flex-end', position: 'relative' }}>
      <Avatar
        alt="Avatar"
        src="https://xsgames.co/randomusers/assets/avatars/female/11.jpg"
        sx={{ alignSelf: 'flex-end', margin: '10px 20px 5px 0' }}
      />
      <Card
        sx={{
          width: '100%',
          background: 'background.paper',
          position: 'relative',
          padding: '16px',
          boxShadow: 'none',
          overflow: 'visible',
          '&::after': {
            content: '""',
            position: 'absolute',
            top: '92%',
            left: '-10px',
            width: 0,
            height: 0,
            borderTop: '10px solid transparent',
            borderBottom: '10px solid transparent',
            borderRight: '10px solid #1c1c1c ',
            transform: 'translateY(-50%)',
          },
        }}
      >
        <form onSubmit={handleSubmit} style={{ display: 'flex', flexDirection: 'column' }}>
          <TextField
            label="Your name"
            value={post.author}
            onChange={(e) => setPost({ ...post, author: e.target.value })}
            // required
            sx={{ marginBottom: '16px' }}
            color="secondary"
            error={!!errors.author}
            helperText={errors.author}
          />
          <TextField
            label="What's on your mind?"
            value={post.body}
            onChange={(e) => setPost({ ...post, body: e.target.value })}
            // required
            multiline
            rows={5}
            sx={{ marginBottom: '16px' }}
            color="secondary"
            error={!!errors.body}
            helperText={errors.body}
          />
          <Grid
            container
            sx={{ width: 'fit-content', marginLeft: 'auto', alignItems: 'center', gap: '1rem', cursor: 'pointer' }}
          >
            <AddAPhotoTwoToneIcon />
            <AddPhotoAlternateTwoToneIcon />
            <AttachFileIcon />
            <Button type="submit" variant="contained" color="primary">
              Create Post
            </Button>
          </Grid>
        </form>
      </Card>
    </Box>
  );
};

export default CreatePost;
