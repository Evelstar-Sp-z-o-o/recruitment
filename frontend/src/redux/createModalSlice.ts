import { createSlice } from '@reduxjs/toolkit';

const createModalSlice = createSlice({
  name: 'createModal',
  initialState: { isOpen: false },
  reducers: {
    openModal: (state) => {
      state.isOpen = true;
    },
    closeModal: (state) => {
      state.isOpen = false;
    },
  },
});

export const { openModal, closeModal } = createModalSlice.actions;
export default createModalSlice.reducer;
