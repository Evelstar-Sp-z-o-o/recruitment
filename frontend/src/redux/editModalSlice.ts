import { createSlice } from '@reduxjs/toolkit';

const editModalSlice = createSlice({
  name: 'editModal',
  initialState: { isOpen: false },
  reducers: {
    openModal: (state) => {
      state.isOpen = true;
    },
    closeModal: (state) => {
      state.isOpen = false;
    },
  },
});

export const { openModal, closeModal } = editModalSlice.actions;
export default editModalSlice.reducer;
