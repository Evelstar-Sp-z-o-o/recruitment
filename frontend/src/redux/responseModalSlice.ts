import { createSlice } from '@reduxjs/toolkit';

const initialState = { isOpen: false, content: '' };

const responseModalSlice = createSlice({
  name: 'responseModal',
  initialState,
  reducers: {
    openModal: (state, action) => {
      state.content = action.payload;
      state.isOpen = true;
    },
    closeModal: (state) => {
      state.isOpen = false;
    },
  },
});

export const { openModal, closeModal } = responseModalSlice.actions;

export default responseModalSlice.reducer;
