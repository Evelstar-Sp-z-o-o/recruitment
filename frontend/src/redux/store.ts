import { configureStore } from '@reduxjs/toolkit';

import responseModalReducer from './responseModalSlice';

const store = configureStore({
  reducer: {
    responseModal: responseModalReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
