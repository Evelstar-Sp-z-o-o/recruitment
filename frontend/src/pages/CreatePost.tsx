import { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';

import PostFormModal from '../components/modals/PostFormModal';
import PostResponseModal from '../components/modals/PostResponseModal';
import { closeModal } from '../redux/createModalSlice';
import { RootState } from '../redux/store';

const CreatePost = () => {
  const [responseModal, setResponseModal] = useState(false);
  const [responseMessage, setResponseMessage] = useState<string>();
  const createModal = useSelector((state: RootState) => state.createModal.isOpen);
  const dispatch = useDispatch();

  const navigate = useNavigate();

  const handleSubmit = async (newPost) => {
    try {
      const response = await fetch('/api/posts', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ data: newPost }),
      });

      if (!response.ok) {
        dispatch(closeModal());
        setResponseMessage('Failed to create a post!');
        setResponseModal(true);
        return;
      }

      const data = await response.json();
      console.log(data);

      dispatch(closeModal());
      setResponseMessage('Successfully created a post!');
      setResponseModal(true);
    } catch (error) {
      console.log(error);
    }
  };

  const handleClose = () => {
    dispatch(closeModal());
    navigate(-1);
  };

  return (
    <>
      <PostFormModal onClose={handleClose} onSubmit={handleSubmit} type="create" isOpen={createModal} />
      <PostResponseModal
        open={responseModal}
        onClose={() => {
          setResponseModal(false);
          navigate('/posts');
        }}
        content={responseMessage}
      />
    </>
  );
};

export default CreatePost;
