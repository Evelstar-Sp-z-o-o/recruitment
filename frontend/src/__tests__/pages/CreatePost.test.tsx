import { Provider } from 'react-redux';
import { MemoryRouter, Route, Routes } from 'react-router-dom';

import { vi } from 'vitest';

import CreatePost from '@/src/pages/CreatePost';
import { openModal } from '@/src/redux/responseModalSlice';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { store } from '../App.test';

// Mocking useLocation
vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom');
  return {
    ...actual,
    useLocation: () => ({ pathname: '/create' }),
    useNavigate: () => vi.fn(),
  };
});

describe('Post creation flow', () => {
  beforeEach(() => {
    vi.spyOn(store, 'dispatch');
  });
  test('initializes elements when modal is opened', async () => {
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/create']}>
          <Routes>
            <Route path="/create" element={<CreatePost refetchPosts={vi.fn()} />} />
          </Routes>
        </MemoryRouter>
      </Provider>,
    );

    await waitFor(() => {
      const textInput = screen.queryByRole('textbox');
      // Check if textarea exists with empty box
      expect(textInput).toBeInTheDocument();
      expect(textInput).toHaveTextContent('');
    });
  });

  test('alert appears when empty string is submitted', async () => {
    const user = userEvent.setup();
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/create']}>
          <Routes>
            <Route path="/create" element={<CreatePost refetchPosts={vi.fn()} />} />
          </Routes>
        </MemoryRouter>
      </Provider>,
    );

    const textInput = screen.getByRole('textbox');
    const submitButton = screen.getByRole('button', { name: /submit/i });

    // Input empty string
    await user.clear(textInput);
    await user.type(textInput, ' ');

    // Click submit button
    await user.click(submitButton);

    //Check if alert appears
    await waitFor(() => {
      const alert = screen.getByTestId('SuccessOutlinedIcon');
      expect(alert).toBeInTheDocument();
    });

    // Input strings in textarea
    await user.clear(textInput);
    await user.type(textInput, 'post');

    // Click submit button
    await user.click(submitButton);

    // Check if alert disappears
    await waitFor(() => {
      const alert = screen.queryByTestId('SuccessOutlinedIcon');
      expect(alert).not.toBeInTheDocument();
    });
  });

  test('dispatches openModal action with failure message when creation fails', async () => {
    // Mocking fetch
    global.fetch = vi.fn().mockResolvedValue({
      ok: false,
    });

    const user = userEvent.setup();

    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/create']}>
          <Routes>
            <Route path="/create" element={<CreatePost refetchPosts={vi.fn()} />} />
          </Routes>
        </MemoryRouter>
      </Provider>,
    );

    const textInput = screen.queryByRole('textbox');
    const submitButton = screen.getByRole('button', { name: /submit/i });

    // Input new post in textarea
    await user.clear(textInput);
    await user.type(textInput, 'new post');

    // Click submit button
    await user.click(submitButton);

    // Check if error message is opened when creation fails
    await waitFor(() => {
      expect(store.dispatch).toHaveBeenCalledWith(openModal('Failed to create a post!'));
    });
  });

  test('dispatches openModal action with success message when creation is successful', async () => {
    // Mocking fetch
    global.fetch = vi.fn().mockResolvedValue({
      ok: true,
    });

    const user = userEvent.setup();

    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/create']}>
          <Routes>
            <Route path="/create" element={<CreatePost refetchPosts={vi.fn()} />} />
          </Routes>
        </MemoryRouter>
      </Provider>,
    );

    const textInput = screen.queryByRole('textbox');
    const submitButton = screen.getByRole('button', { name: /submit/i });

    // Input new post in textarea
    await user.clear(textInput);
    await user.type(textInput, 'new post');

    // Click submit button
    await user.click(submitButton);

    // Check if error message is opened when creation fails
    await waitFor(() => {
      expect(store.dispatch).toHaveBeenCalledWith(openModal('Successfully created a post!'));
    });
  });
});
