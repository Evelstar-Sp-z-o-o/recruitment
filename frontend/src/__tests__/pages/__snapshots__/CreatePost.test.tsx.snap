// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > should match snapshot 1`] = `
<div
  className="main"
>
  <div
    className="sidebar-menu"
  >
    <button
      aria-label="menu"
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeMedium css-1d6w9lk-MuiButtonBase-root-MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-bq05nv-MuiSvgIcon-root"
        data-testid="MenuIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 18h18v-2H3zm0-5h18v-2H3zm0-7v2h18V6z"
        />
      </svg>
    </button>
    <div
      className="logo-sm-box"
    >
      <img
        alt="logo"
        className="logo-sm-image"
        src="/assets/logo-sm.svg"
      />
      <span
        className="page-name"
      >
        Home
      </span>
    </div>
  </div>
  <ul
    className="MuiList-root MuiList-padding sidebar-list css-h4y409-MuiList-root"
  >
    <div
      className="logo-box"
    >
      <img
        alt="logo"
        className="logo-image"
        src="/assets/logo.svg"
      />
    </div>
    <div
      aria-label="menu"
      className="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters Mui-selected MuiListItemButton-root MuiListItemButton-gutters Mui-selected css-1iksp6b-MuiButtonBase-root-MuiListItemButton-root"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="HomeIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
          />
        </svg>
      </div>
      <div
        className="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Home
        </span>
      </div>
    </div>
    <div
      className="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-1yp53vg-MuiButtonBase-root-MuiListItemButton-root"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="NoteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="m22 10-6-6H4c-1.1 0-2 .9-2 2v12.01c0 1.1.9 1.99 2 1.99l16-.01c1.1 0 2-.89 2-1.99zm-7-4.5 5.5 5.5H15z"
          />
        </svg>
      </div>
      <div
        className="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          My Posts
        </span>
      </div>
    </div>
    <a
      href="/create"
      onClick={[Function]}
      style={
        {
          "textDecoration": "none",
        }
      }
    >
      <div
        className="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-9dhvxx-MuiButtonBase-root-MuiListItemButton-root"
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={0}
      >
        <div
          className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1d159sf-MuiSvgIcon-root"
            data-testid="PostAddIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M17 19.22H5V7h7V5H5c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2v-7h-2z"
            />
            <path
              d="M19 2h-2v3h-3c.01.01 0 2 0 2h3v2.99c.01.01 2 0 2 0V7h3V5h-3zM7 9h8v2H7zm0 3v2h8v-2h-3zm0 3h8v2H7z"
            />
          </svg>
        </div>
        <div
          className="MuiListItemText-root css-tlelie-MuiListItemText-root"
        >
          <span
            className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Create a post
          </span>
        </div>
      </div>
    </a>
  </ul>
  <div
    className="MuiBox-root css-zlelhh"
  >
    <div
      className="MuiBox-root css-159nmtr"
    >
      <div
        className="MuiBox-root css-1gpjax4"
        data-testid="latest-box"
        onClick={[Function]}
      >
        Latest
      </div>
      <div
        className="MuiBox-root css-b66kpg"
        data-testid="popular-box"
        onClick={[Function]}
      >
        Most popular
      </div>
    </div>
    <p
      className="MuiTypography-root MuiTypography-body2 css-7005vn-MuiTypography-root"
    >
      There are no posts yet
    </p>
  </div>
</div>
`;
